File: CourseMetrics.jl
module CourseMetrics

include("Output.jl")
include("Utils.jl")

import CurricularAnalytics: blocking_factor, centrality, complexity, Curriculum, delay_factor
import .Output: output, plans, termname
import .Utils: convert, writerow

open("./files/courses_fa12.csv", "w") do file
  writerow(file, [
    "Year",
    "Major",
    "Course",
    "Complexity",
    "Centrality",
    "Year taken in plan",
    "Blocking factor",
    "Delay factor",
  ])

  for year in 2015:2050
    if year ∉ keys(plans)
      break
    end
    for major in sort(collect(keys(plans[year])))
      degree_plans = output(year, major)
      curriculum_college = first(
        college
        for college in ["TH", "WA", "SN", "MU", "FI", "RE", "SI", "EI"]
        if college ∈ keys(degree_plans) && !(college == "SN" && year < 2020)
      )
      curriculum = convert(Curriculum, degree_plans[curriculum_college])
      for (i, course) in enumerate(curriculum.courses)
        if course.prefix == "" || course.institution != "DEPARTMENT"
          continue
        end

        writerow(file, [
          string(year), # Year
          major, # Major
          "$(course.prefix) $(course.num)", # Course
          string(complexity(curriculum, i)), # Complexity
          string(centrality(curriculum, i)), # Centrality
          # Use the curriculum college's year, even though it could vary between
          # colleges
          string((findfirst(course ∈ term.courses for term in degree_plans[curriculum_college].terms) - 1) ÷ 3 + 1),
          string(blocking_factor(curriculum, i)), # Blocking factor
          string(delay_factor(curriculum, i)), # Delay factor
        ])
      end
    end
  end
end

end

File: course_metrics.py
"""
python3 course_metrics.py
"""

from curricularanalytics import Course

from output import MajorOutput
from parse import major_plans
from util import CsvWriter, float_str

HEADER = [
    "Year",
    "Major",
    "Course",
    "Complexity",
    "Centrality",
    "Year taken in plan",
    "Blocking factor",
    "Delay factor",
]


def main() -> None:
    with open("./files/courses_fa12_py.csv", "w") as file:
        writer = CsvWriter(len(HEADER), file)
        writer.row(*HEADER)

        for year in range(2015, 2050):
            majors = major_plans(year)
            if majors == {}:
                break
            for major, plans in majors.items():
                degree_plan = MajorOutput(plans).output_degree_plan()
                curriculum = degree_plan.curriculum
                for course in curriculum.courses:
                    assert isinstance(course, Course)
                    if course.prefix == "":
                        continue
                    writer.row(
                        str(year),  # Year
                        major,  # Major
                        f"{course.prefix} {course.num}",  # Course
                        float_str(curriculum.complexity(course)),  # Complexity
                        str(curriculum.centrality(course)),  # Centrality
                        str(degree_plan.find_term(course)),  # Year taken in plan
                        float_str(
                            curriculum.blocking_factor(course)
                        ),  # Blocking factor
                        float_str(curriculum.delay_factor(course)),  # Delay factor
                    )


if __name__ == "__main__":
    main()
